src/workerd/api/sockets.h

```diff
@@ -46,6 +50,11 @@ class Socket: public jsg::Object {
   jsg::Promise<void> close(jsg::Lock& js);
   // Closes the socket connection.
 
+  jsg::Ref<Socket> startTls(jsg::Lock& js, jsg::Optional<kj::String> expectedServerHostname);
```

Comment:
Is there a plan yet to support ALPN? Would that be configured via the binding? I'm just wondering if the argument here shouldn't be an options struct that would allow us to add more options later without changing the basic signature here.

```cpp
struct StartTlsOptions {
  jsg::Optional<kj::String> expectedServerHostname;
  JSG_STRUCT(expectedServerHostname);
};

jsg::Ref<Socket> startTls(jsg::Lock& js, jsg::Optional<StartTlsOptions> options);
```

Also, quick clarifying question: how is the `expectedServerHostname` here expected to be used? Is it setting the SNI extension on the TLS client handshake or is it used to just verify whatever the server specifies in its response?

Reply:
Good point on making this a struct.

I haven't thought about ALPN yet, mainly followed the option that kj provides. By default the code will use the domain name that you are connecting to, but I figured it would be nice to give the user the ability to override this. The KJ code for this is here: https://github.com/capnproto/capnproto/blob/master/c%2B%2B/src/kj/compat/tls.c%2B%2B#L194. I believe this is a peer certificate check.

Reply:
I definitely think we need to be carefully considering both SNI and ALPN as options. SNI is critical for multi-homing and ALPN is becoming more and more important for protocol negotiation. Let's just make sure we have a plan to address both.

Reply:
Yeah, I think having a struct here should cover us. But let me know if I'm missing something.